geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/stats/GeoMesaMetadataStats.scala

```diff
@@ -0,0 +1,465 @@
+      env.expandToInclude(bounds.upper.getEnvelopeInternal)
+      filterBounds.intersection(env)
+    }.getOrElse(filterBounds)
+  }
+
+  override def getAttributeBounds[T](sft: SimpleFeatureType,
+                                     attribute: String,
+                                     filter: Filter,
+                                     exact: Boolean): Option[AttributeBounds[T]] = {
+    val stat = if (exact) {
+      runStats[MinMax[T]](sft, Stat.MinMax(attribute), filter).headOption
+    } else {
+      readStat[MinMax[T]](sft, GeoMesaMetadataStats.minMaxKey(attribute))
+    }
+    stat.filterNot(_.isEmpty).map(s => AttributeBounds(s.min, s.max, s.cardinality))
+  }
+
+  override def getStats[T <: Stat](sft: SimpleFeatureType,
+                                   attributes: Seq[String],
+                                   options: Seq[Any])(implicit ct: ClassTag[T]): Seq[T] = {
+    val toRetrieve = {
+      val available = GeoMesaStats.statAttributesFor(sft)
+      if (attributes.isEmpty) available else attributes.filter(available.contains)
+    }
+
+    val clas = ct.runtimeClass
+
+    val stats = if (clas == classOf[CountStat]) {
+      readStat[CountStat](sft, countKey()).toSeq
+    } else if (clas == classOf[MinMax[_]]) {
+      toRetrieve.flatMap(a => readStat[MinMax[Any]](sft, minMaxKey(a)))
+    } else if (clas == classOf[Histogram[_]]) {
+      toRetrieve.flatMap(a => readStat[Histogram[Any]](sft, histogramKey(a)))
+    } else if (clas == classOf[Frequency[_]]) {
+      if (options.nonEmpty) {
+        // we are retrieving the frequency by week
+        val weeks = options.map(_.asInstanceOf[Short])
+        val frequencies = toRetrieve.flatMap { a =>
+          weeks.map(frequencyKey(a, _)).flatMap(readStat[Frequency[Any]](sft, _))
+        }
+        Frequency.combine(frequencies).toSeq
+      } else {
+        toRetrieve.flatMap(a => readStat[Frequency[Any]](sft, frequencyKey(a)))
+      }
+    } else if (clas == classOf[Z3Histogram]) {
+      val z = for {
+        geom <- Option(sft.getGeomField)
+        dtg  <- sft.getDtgField
+        if toRetrieve.contains(geom) && toRetrieve.contains(dtg)
+      } yield {
+        // z3 histograms are stored by week - calculate the weeks to retrieve
+        // either use the options if passed in, or else calculate from the time bounds
+        val weeks: Seq[Short] = if (options.nonEmpty) { options.map(_.asInstanceOf[Short]) } else {
+          readStat[MinMax[Date]](sft, minMaxKey(dtg)).map { bounds =>
+            val lt = Z3Table.getWeekAndSeconds(bounds.min.getTime)._1
+            val ut = Z3Table.getWeekAndSeconds(bounds.max.getTime)._1
+            Range.inclusive(lt, ut).map(_.toShort)
+          }.getOrElse(Seq.empty)
+        }
+        val histograms = weeks.map(histogramKey(geom, dtg, _)).flatMap(readStat[Z3Histogram](sft, _))
+        // combine the week splits into a single stat
+        Z3Histogram.combine(histograms)
+      }
+      z.flatten.toSeq
```

Comment:
did my best... see if it's any clearer. I like the for/yield when needing two options, otherwise you have to do `opt1.flatMap(o1 => opt2.map((_, o1)))` or something...